packages/perf/android/src/main/java/io/invertase/firebase/perf/ScreenTrace.java

```diff
@@ -0,0 +1,227 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import android.app.Activity;
+import android.util.Log;
+import android.util.SparseIntArray;
+import android.view.WindowManager;
+
+import androidx.core.app.FrameMetricsAggregator;
+
+import com.google.firebase.perf.FirebasePerformance;
+import com.google.firebase.perf.metrics.Trace;
+import com.google.firebase.perf.util.Constants;
+
+/**
+ * Utility class to capture Screen rendering information (Slow/Frozen frames) for the
+ * {@code Activity} passed to the constructor {@link ScreenTrace#ScreenTrace(Activity, String)}.
+ * <p>
+ * Learn more at https://firebase.google.com/docs/perf-mon/screen-traces?platform=android.
+ * <p>
+ * A slow screen rendering often leads to a UI Jank which creates a bad user experience. Below are
+ * some tips and references to understand and fix common UI Jank issues:
+ * - https://developer.android.com/topic/performance/vitals/render.html#fixing_jank
+ * - https://youtu.be/CaMTIgxCSqU (Why 60fps?)
+ * - https://youtu.be/HXQhu6qfTVU (Rendering Performance)
+ * - https://youtu.be/1iaHxmfZGGc (Understanding VSYNC)
+ * - https://www.youtube.com/playlist?list=PLOU2XLYxmsIKEOXh5TwZEv89aofHzNCiu (Android Performance Patterns)
+ * <p>
+ * References:
+ * - Fireperf Source Code: https://bityl.co/5v2O
+ */
+public class ScreenTrace {
+
+  private static final String TAG = "RNFirebasePerf";
+  private static final String FRAME_METRICS_AGGREGATOR_CLASSNAME =
+    "androidx.core.app.FrameMetricsAggregator";
+
+  private final Activity activity;
+  private final boolean isScreenTraceSupported;
+  private final String traceName;
+
+  private FrameMetricsAggregator frameMetricsAggregator;
+  private Trace perfScreenTrace;
+
+  /**
+   * Default constructor for this class.
+   *
+   * @param activity for which the screen traces should be recorded.
+   * @param tag      used as an identifier for the name to be used to log screen rendering
+   *                 information (like "MyFancyScreen").
+   * @implNote It will automatically force enable hardware acceleration for the passed {@code activity}.
+   * @see #enableHardwareAcceleration(Activity)
+   */
+  public ScreenTrace(Activity activity, String tag) {
+    this.activity = activity;
+    // We don't care about adding the activity name to the trace name
+    // because RN doesn't care about activities
+    this.traceName = tag;
+
+    enableHardwareAcceleration(activity);
+
+    isScreenTraceSupported = isScreenTraceSupported(activity);
+
+    if (isScreenTraceSupported) {
+      frameMetricsAggregator = new FrameMetricsAggregator();
+    }
+  }
+
+  // region Public APIs
+
+  /**
+   * Force enable Hardware acceleration to support screen traces as we can't observe frame
```

Comment:
Ooo! This is dangerous. I have some deep scar tissue from android hardware acceleration. This side effect needs to be clearly documented (https://github.com/ankidroid/Anki-Android/issues/7369#issuecomment-794410275 / https://github.com/ankidroid/Anki-Android/issues/8160).

I am okay with it going in only with a warning that doing a perf trace requires hardware acceleration, which has caused problems in the past, may cause them in the future, and because of known incompatibilities with Android 8/8.1 plus WebViews it will never be enabled on Android 8/8.1 devices

There are no known issues with any other devices + versions but that's a big enough unexpected problem that I think the only safe way to do this is to check for SDK_INT 26/27 and refuse to enable hardware acceleration.

Later on in the flow this will reject the promise when the trace start is attempted since isScreenTraceSupported will return false

Reply:
fixed